### ISSUE TITLE
useOnResize should ignore visibility changes

### ISSUE THREAD

**Issue by haynesjm42** [xh/hoist-react#1578]
Our `useOnResize` hook will current fire the `fn` when the element goes invisible (width and height both become 0px) which can cause issues for components using it to trigger resizes on child elements. This was causing issues with Charts where the resize on hide was causing performance issues, and it is also causing issues in the new DashContainer where the Golden Layouts element size is not yet fit to its container when it becomes visible again, causing some jankiness in the UI.

I propose that `useOnResize` should detect when the component is becoming invisible, and not fire the `fn`. I can't think of a case where your resize handling would have the exact same logic when the element is hidden/shown, we would likely end up adding a check in each handler to ignore the resize logic when the element is hidden or shown. It would also be nice if the `useOnResize` could somehow not fire the `fn` if the size hasn't actually changed (we have a check in Chart to ignore the resize if the size is the same as our last known size).

We could consider adding a `useOnVisible` or similar hook for components which want to handle visibility changes.

**Comment by peterjcoffman** [comment]
WIP PR is here: https://github.com/xh/hoist-react/pull/1579
I have tested this on charts, but I don't know what to look for for the DashContainer.

> It would also be nice if the useOnResize could somehow not fire the fn if the size hasn't actually changed (we have a check in Chart to ignore the resize if the size is the same as our last known size).

We do that check in Chart only because we ignore resizing to 0 (i.e. hiding). I don't believe a resize event will fire if the size doesn't actually change. But on Chart, a resize *from* zero will have no effect because we ignored the resize *to* zero.
